src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs

```diff
@@ -0,0 +1,196 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+using System.Globalization;
+
+namespace System.Reflection.Emit
+{
+    internal sealed class TypeBuilderImpl : TypeBuilder
+    {
+        internal List<MethodBuilderImpl> _methodDefStore = new();
+        internal List<FieldBuilderImpl> _fieldDefStore = new();
+        private readonly ModuleBuilderImpl _module;
+        private readonly string _name;
+        private readonly string? _namespace;
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)]
+        private Type? _typeParent;
+        private TypeAttributes _attributes;
+
+        internal TypeBuilderImpl(string fullName, TypeAttributes typeAttributes,
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] Type? parent, ModuleBuilderImpl module)
+        {
+            _name = fullName;
+            _module = module;
+            _attributes = typeAttributes;
+            SetParent(parent);
+            // Extract namespace from fullName
+            int idx = _name.LastIndexOf('.');
+
+            if (idx != -1)
+            {
+                _namespace = _name[..idx];
+                _name = _name[(idx + 1)..];
+            }
+        }
+
+        internal ModuleBuilderImpl GetModuleBuilder() => _module;
+        // TODO: Placeholder for "typeof(object)", we don't want to emit reference to runtime System.Private.CoreLib
```

Comment:
note this also applies to primitive types (int, string, etc) not just System.Object.

Reply:
Good point, as core assembly is going to be provided in the factory method, updated to use that core Assembly to load types by name